### PR TITLE
RFC: Identity Resolution

### PR DESCRIPTION
**Pull request by kim**
[Rendered](https://github.com/radicle-dev/radicle-link/blob/rfc/identity-resolution/docs/rfc/identity_resolution.md)

### CODE REVIEW

File: docs/rfc/identity_resolution.md [hunk 1]
```diff
@@ -0,0 +1,249 @@
...
+                |   |-- id # <-- points to the identity document history
+                |   |-- moi # <-- points to the identity of $PEER_ID
+                |   `-- ids
+                |       |-- $CERTIFIER_IDENTITY_1
+                |       `-- $CERTIFIER_IDENTITY_2
+                `-- remotes
+                    |-- $OTHER_PEER_ID
+                    |   `-- refs
+                    |       `-- heads
+                    `-- rad
```
Comment (FintanH) [inline docs/rfc/identity_resolution.md:80]: Expected this to be under `refs` like above. Is it not referring to the `$OTHER_PEER_ID`'s id and certifiers?

File: docs/rfc/identity_resolution.md [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+config it reduces to:
+
+    [include]
+        path = /path/to/managed.inc
+
+Note that we also need to decide on the `HEAD` (ie. default branch to check
+out), but since this is subject to workflow preferences, and dependent on the
+verification result, a discussion is outside the scope of this document.
+
+## Alternative Appraches
```
Comment (FintanH) [inline docs/rfc/identity_resolution.md:196]:
```suggestion
## Alternative Approaches
```

File: docs/rfc/identity_resolution.md [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+
+            for certifier in certifiers()
+                refs/namespaces/$certifier/refs/remotes/$peer/rad/id*:refs/namespaces/$certifier/refs/remotes/$peer/rad/id*
+            end
+        end
+    end
+
+We can now, in a single packfile, receive a "mirror" of the logical remote
+repository requested (modulo the mapping of remotely owned branches to
+`refs/remotes`), _as well as_ of the top-level logical repositories of all
```
Comment (FintanH) [inline docs/rfc/identity_resolution.md:148]:
```suggestion
`refs/remotes`), _as well as_ all of the top-level logical repositories of all
```

File: docs/rfc/identity_resolution.md [hunk 4]
```diff
@@ -0,0 +1,249 @@
...
+verification result, a discussion is outside the scope of this document.
+
+## Alternative Appraches
+
+A similar effect, even exposing the same namespacing scheme on the git protocol
+level, could be achieved by leaving top-level repositories standalone, but
+mutually linking their object databases via [alternates]. An advantage would be
+potentially more efficient [repack]s and resulting packfile layout.
+Disadvantages include handling of symbolic refs, which would require
+filesystem-level symblic links, or a custom `refdb`, and handling of repository
```
Comment (FintanH) [inline docs/rfc/identity_resolution.md:203]:
```suggestion
filesystem-level symbolic links, or a custom `refdb`, and handling of repository
```

File: docs/rfc/identity_resolution.md [hunk 5]
```diff
@@ -0,0 +1,249 @@
...
+deletion, which would require keeping track of a refcount, and prevent removal
+before it has reached zero.
+
+## Drawbacks
+
+* The use of symrefs _below_ the `refs` hierarchy is somewhat unorthodox. As
+  symrefs where invented to replace actual filesystem symbolic links (which are
```
Comment (FintanH) [inline docs/rfc/identity_resolution.md:210]:
```suggestion
  symrefs were invented to replace actual filesystem symbolic links (which are
```

File: docs/rfc/identity_resolution.md [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+  feasible to replace the object and refs database backends entirely.
+
+## Conclusion
+
+Overall, the risks seem manageable, and the reduced complexity for obtaining,
+updating, and verifying `radicle-link`-enabled git repositories appear to
+outweigh them.
+
+As Google and Facebook knew already, all source control problems can be solved
+by a monorepo, and they can't be wrong, can they?
```
Comment (FintanH) [inline docs/rfc/identity_resolution.md:235]: When have they ever been

File: docs/rfc/identity_resolution.md [hunk 7]
```diff
@@ -0,0 +1,249 @@
...
+
+    refs/namespaces/foobar/refs/heads/master
+
+The namespace is accessed by setting the `$GIT_NAMESPACE` environment variable,
+or passing the `--namespace` parameter to the `git` executable. However, the
+only commands which respect this are [git-upload-pack], [git-receive-pack], and
+[git-http-backend]. When invoked namespaced, they will simply only consider the
+`refs` hierarchy below the namespace (`foobar` in the example).
+
+The fact that no other git command seems to be aware of namespaces ([git-gc]
```
[the remark this answers is not in the document]
Comment (kim) [reply docs/rfc/identity_resolution.md:51]: That's the "perhaps surprising" part: I was convinced that "heads", "tags", and "remotes" were hardcoded in some places when it comes to (re-)packing. Apparently, that's wrong: `git-gc` will consider everything below `refs/`. To be safe, we should still use `namespaces`, as that's already established in the git codebase, and should it receive special treatment in the future, that will certainly preserve the current semantics.

File: docs/rfc/identity_resolution.md (hunk 1 above)
Comment (kim) [reply docs/rfc/identity_resolution.md]: Ya my ASCII art failed me

File: docs/rfc/identity_resolution.md (hunk 6 above)
Comment (kim) [reply docs/rfc/identity_resolution.md:235]: 🎸 🕊️ 🎶

**Comment by kim** [comment]
https://github.com/radicle-dev/radicle-link/issues/139#issuecomment-627569602

**Pull request merged**